Add support for short signatures in autosummary

timhoffm opened the pull request:
This introduces a new `:signatures:` option to `.. autosummary::`.

Possible values are:

- "long" (*default*) - same as the current default
- "short": format functions and classes as `func()` if they do not have arguments and `func(…)` if they do have arguments.
- "none": same as the already existing :no-signature:

The new functionality here is the "short" option. This allows to distinguish properties and attributes from functions and classes without costing a lot of space for a long signature.

*Note 1: Design:* I've added the  `:signatures:` option to have a single config value instead of `:shortsignatures:` alongside  `:no-signatures:`. This is also more extensible in case other signature variants come up in the future.  `:no-signatures:` remains available for backward compatibility, but `:signatures:` takes precedence if given.

*Note 2: Tests:* There are currently no tests for the signature formatting. I could add some tests for `Autosummary.get_items`, but I don't understand how to set up everything in a test so that I can exercise that function. Please give a hint how to do this if tests are desired.

Closes #13101.

timhoffm committed:
Move signature option handling into get_items()

get_items() already formats the signature. Handling :nosignatures: option in the saves the signature computation cost if the signature is not wanted. Also, this prepares for more complex signature options.

timhoffm committed:
Add :signatures: option to autosummary directive

This is in preparation of extending singature formating.

timhoffm committed:
Add ":signatures: short" to autosummary directive

This formats functions and classes with (…) if they have
arguments and () if they don't have arguments. This makes
it easier to distinguish callables from attributes and
properties.

timhoffm committed:
Merge branch 'master' into autosummary-short-signature

timhoffm committed:
Merge branch 'master' into autosummary-short-signature

timhoffm committed:
Fix ruff

AA-Turner reviewed (commented):
It would be nice to have tests, but I'm also unfamiliar with `autosummary`, I'd be starting from scratch. Other than that this looks good! 

Please can you update it in light of #13264?

A

timhoffm committed:
Merge branch 'master' into autosummary-short-signature

# Conflicts:
#	doc/usage/extensions/autosummary.rst
#	sphinx/ext/autosummary/__init__.py

timhoffm committed:
Rollback renaming :nosignatures: to :no-signatures:

Since we introduced the more general ``:signatures: none` spelling,
we can directly push people in this direction.

timhoffm committed:
Merge branch 'master' into autosummary-short-signature

AA-Turner reviewed (approved):
Needs a CHANGES entry.

A

timhoffm committed:
Add CHANGES entry

AA-Turner commented on CHANGES.rst:
```diff
@@ -47,10 +47,10 @@ Features added
   warning sub-type.
   Patch by Slawek Figiel.
-* #9732: Add the new ``autodoc.mocked_object`` warnings sub-type.
+* #9732: Add the new ``autodoc.mock_objects`` warnings sub-type.
```
?

timhoffm replied:
Strange. That's of course unintended.. I possibly  have messed up git when bringing in the remote upstream merge.

timhoffm replied:
Ok fixed. I'm struggling with the merge-main-into-branch workflow. I'm much more comfortable with rebases. Sorry for the confusion.

AA-Turner replied:
No worries -- I tend to use the merge interface in PyCharm for tricky merges, which simplifies things greatly.

timhoffm committed:
Reverted unintended edits in CHANGES

timhoffm committed:
Merge branch 'master' into autosummary-short-signature

AA-Turner commented on CHANGES.rst:
```diff
@@ -80,6 +80,8 @@ Features added
   :confval:`autodoc_default_options`.
   Patch by Jonny Saunders and Adam Turner.
+* #13172: Add support for short signatures in autosummary.
+  Patch by Tim Hoffmann
```
```suggestion
  Patch by Tim Hoffmann.
```

AA-Turner committed:
Full stop.
